src/neuralnet/voting/builders.cpp

```diff
@@ -0,0 +1,1142 @@
+    }; // PubkeyDestinationFilter
+
+    //!
+    //! \brief Get a set of filtered outputs spent for public key addresses.
+    //!
+    //! \param outputs The set of outputs to filter. Sorted in descending order
+    //! by amount.
+    //!
+    //! \return Filtered outputs grouped by address.
+    //!
+    static OutputAddressMap FilterOutputs(const std::vector<COutput>& outputs)
+    {
+        OutputAddressMap by_address;
+        PubkeyDestinationFilter pubkey_filter(by_address);
+
+        for (const auto& txo : outputs) {
+            CTxDestination dest;
+
+            if (!ExtractDestination(txo.tx->vout[txo.i].scriptPubKey, dest)) {
+                LogPrint(LogFlags::VOTE, "  skipped invalid output in %s",
+                    txo.tx->GetHash().ToString());
+                continue;
+            }
+
+            // Avoid voting with outputs less than 1 GRC. This avoids bloating
+            // the vote contract with insignificant weight that costs the user
+            // additional fees:
+            //
+            if (txo.tx->vout[txo.i].nValue < COIN) {
+                LogPrint(LogFlags::VOTE, "  skipped < 1 GRC outputs for: %s",
+                    DestinationToAddressString(dest));
+
+                break; // The input set is sorted so the rest are too small.
+            }
+
+            // Check that the output is P2PK or P2PKH. Voting does not support
+            // more advanced redemption scripts like multisig yet:
+            //
+            auto filter = std::bind(pubkey_filter, std::placeholders::_1, txo);
+            boost::apply_visitor(filter, dest);
+        }
+
+        return by_address;
+    }
+
+    //!
+    //! \brief Get an ordered set of output address groups.
+    //!
+    //! \param by_address The set of outputs grouped by address to sort.
+    //!
+    //! \return Output address groups sorted by the total amount of the outputs
+    //! in each group in descending order.
+    //!
+    static OutputAddressSet SortAddressesByAmount(OutputAddressMap by_address)
+    {
+        OutputAddressSet ordered;
+        ordered.reserve(by_address.size());
+
+        for (auto& address_pair : by_address) {
+            ordered.emplace_back(std::move(address_pair.second));
+        }
+
+        // Order the output address groups by total amount. For wallets with
+        // more outputs than the number of outputs that will fit in a voting
+        // transaction, we want to choose the largest addresses first, so we
+        // sort in descending order.
+        //
+        // TODO: Ordering by address is not perfect because an address could
+        // contain one large output and hundreds of small outputs. We should
+        // use a more accurate order based on something like the average for
+        // the output amounts associated with the address.
+        //
+        std::sort(ordered.begin(), ordered.end(), std::greater<AddressOutputs>());
```

Comment:
I think a better way to do this would be to pivot the outputs and the addresses and then order by outputs descending. the addresses referred to by the outputs chosen is implicitly the set of addresses that would be included. Note that not all UTXO's would be included for a given address, but that is ok... that is exactly what we want. I haven't thought about the ramifications of my suggestion on other parts of your code.

Reply:
I don't think using the average as an alternative is going to be particularly helpful, because many wallets are going to have a UTXO size that is exponentially distributed, with a size cutoff roughly at 4-5k (which is what the stake splitter for stake optimization generates with the current difficulty).

Reply:
If changing this doesn't break validation code, then optimizing this is a leisure...

Reply:
This is just on the creation side, so it changing the output selection won't impact consensus. You're right about using an average... we really need to trim the outputs before grouping them.

Reply:
Let's see how the existing design works in testing...